heapster-saw/src/Verifier/SAW/Heapster/Implication.hs

```diff
@@ -4023,6 +4045,62 @@ implElimPopIthLLVMBlock x ps i
 implElimPopIthLLVMBlock _ _ _ = error "implElimPopIthLLVMBlock: malformed inputs"
 
 
+-- | Assume the top of the stack contains @x:p1*...*pn@, which are all the
+-- permissions for @x@. Extract the @i@th conjuct @pi@, which should be a
+-- @memblock@ permission. Eliminate that @memblock@ permission using
+-- 'implElimLLVMBlock' if possible to atomic permissions @x:q1*...*qm@, and
+-- append the resulting atomic permissions @qi@ to the top of the stack, leaving
+--
+-- > x:ps1 * ... * pi-1 * pi+1 * ... * pn * q1 * ... * qm
+--
+-- on top of the stack. Return the list of atomic permissions that are now on
+-- top of the stack. If the @memblock@ permission @pi@ cannot be elimnated, then
+-- fail.
+implElimAppendIthLLVMBlock :: (1 <= w, KnownNat w, NuMatchingAny1 r) =>
+                              ExprVar (LLVMPointerType w) ->
+                              [AtomicPerm (LLVMPointerType w)] -> Int ->
+                              ImplM vars s r (ps :> LLVMPointerType w)
+                              (ps :> LLVMPointerType w)
+                              [AtomicPerm (LLVMPointerType w)]
+implElimAppendIthLLVMBlock x ps i
+  | i < length ps , Perm_LLVMBlock bp <- ps!!i =
```

Comment:
Well, I see what you are saying, but all of the `Int`s generated in this whole module are always non-negative, since they are generated by `findIndex`. In fact, I could almost see omitting even this check, since it will call `error` either way if the check fails. Further, all of the calls to this function are structured in such a way as to always generate `i` by calling `findIndex` on `ps`. I dunno, what do you think?

Reply:
It seems like we should check both or neither. The values are all generated from a matching call to `findIndex` now, but it's plausible someone could call this function differently in the future or run findIndex on a mismatched list. I think what we really want is just a checked version of `!!` here.

We're already pulling in lens, so we could just do: `Perm_LLVMBlock bp <- ps ^? ix i`

I'd be OK with removing the check altogether, it just seems wrong have half the check.

Reply:
Ooh, that's some fancy lens stuff. I do agree with you that it seems wrong to have half the check. On top of that, most other functions in this module that take `Int`s don't do any checks, and it seems weird to do them here and not other places as well. So, yeah, I'll just drop it here.